Implement new data component predicates in block predicates

FluxCapacitor2 opened the pull request:
## Proposed changes

Implements data component predicates, which are now included in block predicates. Resolves the issue described in [this comment](https://github.com/Minestom/Minestom/pull/2699#issuecomment-2848516382) and this TODO comment:
https://github.com/Minestom/Minestom/blob/2398778b46cd1718780a62b74c85c9ad9a041363/src/main/java/net/minestom/server/instance/block/predicate/BlockPredicate.java#L56

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Make sure the client can read all the new predicates correctly (test manually)
- [x] Make sure that the `container` predicate works with an item predicate
- [x] Add support for block tags
- [x] Clean up unused network types (IntBounds/DoubleBounds are only ever serialized to NBT, so the network type is unnecessary)
- [ ] Consider renaming some of the classes to be less confusing. The predicates are named the same as some of their data component counterparts, which might cause incorrect IDE suggestions. Should I add a `*Predicate` suffix to all of them?
- [x] Add unit tests to make sure my custom network types can encode a BlockPredicate with data component predicates and then decode it again to the same object
- [x] Add some Javadoc comments

## Remaining questions

1. In `DataComponentPredicates`, I keep a list of the different predicate types in the order that the client registers them in a registry. Is there a way to autogenerate that, and if so, is it worth the effort?
2. Are there any other places that these predicates can/should be used, like advancements or entity predicates in commands?

I'm leaving this PR in draft mode until more of the checklist above is completed. Feel free to leave feedback in the meantime!

## Breaking Changes

- The `minecraft:lock` data component is now an `ItemPredicate` instead of `CustomData`
- The 3-argument `BlockPredicate` constructor which previously accepted a `CompoundBinaryTag` now accepts an `NbtPredicate`. The 1-argument constructor has a wrapper that creates an NbtPredicate from the given CompoundBinaryTag, but the 3-argument constructor doesn't.

FluxCapacitor2 committed:
Remove unused network type

FluxCapacitor2 committed:
Replace `Bounds` with `Range`

GoldenStack commented on src/main/java/net/minestom/server/instance/block/predicate/CollectionPredicate.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    @Override
+    public boolean test(Iterable<T> collection) {
```
From what I can tell, the data components that use this structure (`attribute_modifiers`, `bundle_contents`, `container`, `fireworks`, `writable_book_content`, `written_book_content`) all store the items as lists, so this should be able to be converted to a `List<T>` or `Collection<T>` which means you should be able to work over one of those types and remove the ugly `sizeOf` method and some other associated stuff. LMK if I missed something that makes this impossible, though.

GoldenStack replied:
Okay I can see it used in `AttributeModifiers#test` but I think this isn't at all a hard dependency.

GoldenStack commented on src/main/java/net/minestom/server/instance/block/predicate/CollectionPredicate.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.List;
+import java.util.function.Predicate;
+
+public record CollectionPredicate<T, P extends Predicate<T>>(Contains<T, P> contains, Count<T, P> counts,
```
Should probably at least put `@Nullable` on the nullable fields (you omit the `@Nullable` in other field definitions too but I'm just mentioning it here), and preferably also `@NotNull` on the non-nullable ones.

GoldenStack commented on src/main/java/net/minestom/server/instance/block/predicate/DataComponentPredicate.java:
```diff
@@ -0,0 +1,376 @@
+        }
+    }
+
+    record StoredEnchantment(
+            @Nullable List<DynamicRegistry.Key<net.minestom.server.item.enchant.Enchantment>> enchantments,
+            Range.Int levels) {
+
+        public static Codec<StoredEnchantment> CODEC = StructCodec.struct(
+                "enchantments", net.minestom.server.item.enchant.Enchantment.CODEC.listOrSingle().optional(), StoredEnchantment::enchantments,
+                "levels", DataComponentPredicates.INT_RANGE_CODEC.optional(), StoredEnchantment::levels,
+                StoredEnchantment::new
+        );
+
+        public boolean test(DataComponent.Holder holder) {
+            EnchantmentList holderEnchants = holder.get(DataComponents.STORED_ENCHANTMENTS);
+            if (holderEnchants == null) {
+                return false;
+            }
+            if (enchantments == null) {
+                // If `enchantments` is not specified, the predicate returns true when any enchantment matches the specified `level`
+                return holderEnchants.enchantments().entrySet().stream().anyMatch(entry -> levels == null || levels.inRange(entry.getValue()));
+            }
+            for (DynamicRegistry.Key<net.minestom.server.item.enchant.Enchantment> e : enchantments) {
+                if (holderEnchants.has(e) && (levels == null || levels.inRange(holderEnchants.level(e)))) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
```
You should be able to deduplicate this by having `Enchantments`/`StoredEnchantments` pass in the relevant `EnchantmentList`. There don't need to be two implementations.

GoldenStack commented on src/main/java/net/minestom/server/instance/block/predicate/DataComponentPredicate.java:
```diff
@@ -0,0 +1,376 @@
+    }
+
+    record Potions(List<PotionType> potionTypes) implements DataComponentPredicate {
+        public static Codec<Potions> CODEC = PotionType.CODEC.listOrSingle().transform(Potions::new, Potions::potionTypes);
```
This doesn't handle the case of there being a tag of potion types. Essentially, per [the wiki](https://minecraft.wiki/w/Data_component_predicate#potion_contents) (and I did confirm otherwise that this is how it actually works), you need to be able to support a tag of something, e.g. for items, if the string `#minecraft:logs` is passed in it needs to refer to all logs. You also need to handle this in many other cases across the `DataComponent` predicates but I'm just mentioning this here so I don't have to spam it everywhere.

FluxCapacitor2 replied:
It looks like this isn't possible for the `jukebox_playable` and `trim` predicates until the tag rework PR #2709 is merged, but I added tag support for `enchantments`, `stored_enchantments`, and `potion_contents`.

GoldenStack commented on src/main/java/net/minestom/server/instance/block/predicate/DataComponentPredicate.java:
```diff
@@ -0,0 +1,376 @@
+        }
+    }
+
+    record CustomData(BinaryTag nbt) implements DataComponentPredicate {
```
Just like I mentioned for `BlockPredicate#nbt` this does also need to be a NBT predicate.

GoldenStack commented on src/main/java/net/minestom/server/instance/block/predicate/DataComponentPredicate.java:
```diff
@@ -0,0 +1,376 @@
+        @Override
+        public boolean test(DataComponent.Holder holder) {
+            Iterable<ItemStack> itemStacks = holder.get(DataComponents.CONTAINER);
+            return items().test(itemStacks);
```
This needs to exclude air (or otherwise air needs to always pass/not affect the results).

GoldenStack commented on src/main/java/net/minestom/server/instance/block/predicate/DataComponentPredicate.java:
```diff
@@ -0,0 +1,376 @@
+            ))::iterator);
+        }
+
+        record Entry(Attribute attribute, Key id, Double amount, AttributeOperation operation,
```
Is there a reason why there's this intermediate `Entry` class?

GoldenStack commented on src/main/java/net/minestom/server/instance/block/predicate/DataComponentPredicate.java:
```diff
@@ -0,0 +1,376 @@
+        public boolean test(DataComponent.Holder holder) {
+            var trim = holder.get(DataComponents.TRIM);
+            if (trim == null) return false;
+            if (material != null && material.contains(trim.material().resolve(MinecraftServer.getTrimMaterialRegistry())))
```
This should be `!material.contains`

GoldenStack reviewed (commented):
There are a couple implementation issues but it's definitely mostly correct. LMK if I can help implement anything (since I was going to anyway before you told me about this PR).

FluxCapacitor2 committed:
Fix ArmorTrim predicate returning false when materials match

FluxCapacitor2 committed:
Make CollectionPredicate accept a Collection instead of an Iterable

Not sure why I didn't do this initially

FluxCapacitor2 committed:
Copy lists when necessary to make records truly immutable

FluxCapacitor2 committed:
Ignore air when processing `container` predicate

FluxCapacitor2 committed:
Use existing DataComponent.Holder inherited from ItemStack instead of constructing a new DataComponentMap

FluxCapacitor2 commented on src/main/java/net/minestom/server/item/predicate/ItemPredicate.java:
```diff
@@ -0,0 +1,90 @@
+        if (predicates == null)
+            return true;
+
+        // Construct a DataComponentMap with ALL the item's components, including the default ones originating from the item's material
```
Good idea, I forgot that ItemStack implemented DataComponent.Holder!

FluxCapacitor2 committed:
Fix tests

FluxCapacitor2 commented:
@GoldenStack Thanks for the very thorough review! I have addressed everything you've mentioned and I'll keep working on javadocs, unit tests, and manually verifying that the client reads the all the predicates properly.

I wasn't able to easily add tag support for the jukebox_playable and trim predicates ([thread](https://github.com/Minestom/Minestom/pull/2732#discussion_r2096619807)) because they don't exist in vanilla, so Minestom doesn't load them. It looks like that'll be possible when #2709 lands.

FluxCapacitor2 committed:
Use DataComponentMap's build-in serializer instead of custom one and fix BlockPredicate test with non-null but empty exact or partial matchers

FluxCapacitor2 committed:
Fix serializing data component predicates as NBT

FluxCapacitor2 committed:
Fix decoding data component predicates from NBT

FluxCapacitor2 committed:
Implement new data component predicates in block predicates

FluxCapacitor2 committed:
Encode the minecraft:lock data component as an item predicate

FluxCapacitor2 committed:
Allow tags in enchantment and potion data component predicates

The wiki says that tags should also be allowed in jukebox_playable and trim, but I don't think the game has any tags for those(?)

FluxCapacitor2 committed:
Add nullability annotations

FluxCapacitor2 committed:
Remove duplication of codecs and `test` methods in Enchantments and StoredEnchantments predicates

FluxCapacitor2 committed:
Use NBTPredicate instead of requiring exact tag equality

Thanks GoldenStack for writing the NBT comparison method!

FluxCapacitor2 committed:
Remove unnecessary intermediate `Entry` record in AttributeModifiers predicate

FluxCapacitor2 committed:
Add more convenience constructors, create a CollectionPredicate Builder, and make some erroneously-private classes public

FluxCapacitor2 committed:
Fix testSingleBlockNbtInput

FluxCapacitor2 committed:
Fix demo compile

FluxCapacitor2 commented:
Working on some unit tests now and then I'll be done

FluxCapacitor2 committed:
Add unit tests and Javadoc comments for DataComponentPredicate

mworzala reviewed (commented):
I added some comments but this is overall looking extremely good, thanks for taking this work!

FluxCapacitor2 committed:
Rename componentPredicates -> components for consistency with other names

FluxCapacitor2 committed:
Add more nullability annotations

FluxCapacitor2 committed:
Convert ComponentPredicateType to a registry and simplify Codec and NetworkType implementations

FluxCapacitor2 committed:
Uncomment and fix failing tests

FluxCapacitor2 committed:
Use an enum for component predicate types

FluxCapacitor2 committed:
Remove static reference to registry in favor of MinecraftServer#componentPredicateTypes()

FluxCapacitor2 committed:
Remove reference to deprecated TagStringIOExt

GoldenStack commented:
From testing with loot tables, I've discovered that your EnchantmentListPredicate codec is wrong; it needs to have a default value for the `max` [here](https://github.com/FluxCapacitor2/Minestom/blob/feat/component-predicates/src/main/java/net/minestom/server/instance/block/predicate/DataComponentPredicate.java#L104), since right now 144 (of the 1244 total) vanilla loot tables fail to deserialize because of this. I guess a suitable maximum is probably Byte.MAX_VALUE or Short.MAX_VALUE, but I'm not sure which.

Also when you add the maximum I guess you'd have to make a copy of the int range codec, since it doesn't make sense to have a max value like this outside of enchantments (I'm just saying this because I feel like I might forget if I were doing this).

GoldenStack commented:
I think the Minestom codebase tends to use `@Nullable T` instead of `Optional<T>` so that'd probably be better. And honestly what you said makes a lot of sense, so I'm convinced that's better than what I originally suggested. So I guess I'd just recommend adding`@Nullable T`. Thanks for the detailed response! Hopefully this PR can get merged pretty soon.

FluxCapacitor2 committed:
Specify initial capacity when copying maps

FluxCapacitor2 committed:
Make `min` and `max` in *Range classes optional to allow unbounded ends of ranges

FluxCapacitor2 commented on src/main/java/net/minestom/server/instance/block/predicate/ComponentPredicateSet.java:
```diff
@@ -0,0 +1,80 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public record ComponentPredicateSet(Int2ObjectMap<DataComponentPredicate> predicates) {
```
Do you like that this is immutable? I was looking at DataComponentMap while implementing this

kermandev replied:
A defensive copy should be good here.

FluxCapacitor2 reviewed (commented):
@mworzala I switched ComponentPredicateType to a registry and replaced the type of `DataComponentPredicates#predicates` with a new class, `ComponentPredicateSet`. I like it better than a `Map<ComponentPredicateType, DataComponentPredicate>` because it's typesafe (no risk of key not matching the type of its value) and easier to construct than a Map.

It also gives me a nice place to put the necessary codec and network type. The way I implemented it isn't as optimized as before since it reads them into a Map and then copies it into a `ComponentPredicateSet`, but it's way more readable IMO.

I have a few questions about style below:

FluxCapacitor2 committed:
Add tests for NbtPredicate and fix one inconsistency with vanilla Minecraft's implementation

mworzala commented on src/main/java/net/minestom/server/MinecraftServer.java:
```diff
@@ -309,6 +311,10 @@ public static AdvancementManager getAdvancementManager() {
         return process().enchantmentLocationEffects();
     }
 
+    public static @NotNull DynamicRegistry<Codec<? extends DataComponentPredicate>> componentPredicateTypes() {
+        return process().componentPredicateTypes();
+    }
+
```
So far I have exposed registries like this (ie the enchantment ones are the same). There is an open pr to convert Registries into a Registry of Registry values, in which case they will still be present.

tl;dr: yes

mworzala commented on src/main/java/net/minestom/server/codec/Codec.java:
```diff
@@ -150,11 +156,19 @@ sealed interface RawValue permits CodecImpl.RawValueImpl {
+        return mapValue((ignored) -> valueCodec, Integer.MAX_VALUE);
+    }
+
+    default <V> @NotNull Codec<Map<T, V>> mapValue(@NotNull Function<T, @NotNull Codec<V>> valueCodecGetter, int maxSize) {
+        return new CodecImpl.MapImpl<>(Codec.this, valueCodecGetter, maxSize);
+    }
```
Having the value based on the key screams tagged union (`unionType` or a registry one) to me, not map. I can try to fiddle with it myself soon and see if I understand why its not that

FluxCapacitor2 committed:
Use Java Map type publicly, defensively copy the input map, and add method to remove item

FluxCapacitor2 committed:
Merge branch 'master' into feat/component-predicates

FluxCapacitor2 committed:
Remove some `@NotNull` annotations that I missed

kermandev reviewed (changes requested):
I was mainly focusing on doing a short review on changes involving codecs, but Ive gone ahead and looked through most of it and shared some comments below.

Most of them stem from avoid usage of streams inside test predicates; as they are kind of unnecessary and you seem to already use the alternative in a few places.

For immutable objects you may want to put withX as it could be useful to users, and make sure to add pure contracts for ide warnings.

Also for predicates it seems really bad if everything was null or empty as the predicate would be worthless at that point. Maybe you should throw in the constructor if that happens.

FluxCapacitor2 committed:
Implement various small suggestions from @kermandev

FluxCapacitor2 committed:
Make `exact` and `predicates` non-null in DataComponentPredicates

FluxCapacitor2 committed:
Add javadocs and Contract annotations to new Codec methods

FluxCapacitor2 committed:
ComponentPredicateSet improvements:
- Convert ComponentPredicateSet to a regular (non-record) class to make private new no-copy constructor
- Add javadocs
- Remove unnecessary copy in add()
- Add new convenience constructor from a Collection<DataComponentPredicate>
- Make ComponentPredicateSet implement Predicate<DataComponent.Holder>

kermandev commented on src/main/java/net/minestom/server/instance/block/predicate/ComponentPredicateSet.java:
```diff
@@ -10,19 +11,40 @@
+ * will return a new instance of this class with the element added or removed.
+ *
+ */
+public class ComponentPredicateSet implements Predicate<DataComponent.Holder> {
```
Missing final if this continues to not be a record (see below)

kermandev commented on src/main/java/net/minestom/server/instance/block/predicate/ComponentPredicateSet.java:
```diff
@@ -10,19 +11,40 @@
+
     public ComponentPredicateSet() {
-        this(new Int2ObjectArrayMap<>());
+        this(Int2ObjectMaps.emptyMap());
```
This makes no sense to exist as its equivalent to .EMPTY, if anything it should just be a vararg on the delegate of Collections version.

kermandev commented on src/main/java/net/minestom/server/instance/block/predicate/ComponentPredicateSet.java:
```diff
@@ -10,19 +11,40 @@
+        }
+    }
+
+    private ComponentPredicateSet(Map<Integer, DataComponentPredicate> predicates) {
```
I think its fine to keep this a record, just ensure a `predicates = Int2ObjectMaps.unmodifiable(new Int2ObjectArrayMap<>(predicates))` is performed.

FluxCapacitor2 replied:
I switched it to a regular class because it makes it possible for me to have a private constructor that doesn't copy the input map. This allows the `add` and `remove` methods to only copy the map once instead of twice.

A regular class also makes it possible to completely hide the Map from the API client. This class really should behave more like a Set, and the Map should be considered an implementation detail that's only used during serialization.

Let me know what you think.

kermandev commented on src/main/java/net/minestom/server/instance/block/predicate/ComponentPredicateSet.java:
```diff
@@ -10,19 +11,40 @@
-    public ComponentPredicateSet(Map<Integer, DataComponentPredicate> predicates) {
-        this.predicates = new Int2ObjectArrayMap<>(predicates);
+    public ComponentPredicateSet(Collection<DataComponentPredicate> predicates) {
+        this.predicates = new Int2ObjectArrayMap<>(predicates.size());
```
If your moving this to a record, you can make this implementation into a private static to return a new Int2ObjectArrayMap

kermandev committed:
Merge branch 'master' into fork/FluxCapacitor2/feat/component-predicates

# Conflicts:
#	demo/src/main/java/net/minestom/demo/PlayerInit.java
#	src/main/java/net/minestom/server/codec/Codec.java
#	src/main/java/net/minestom/server/codec/CodecImpl.java

InfiniteCoder01 commented:
What's the current state of this PR? I checked it out with Trove (I am implementing entity predicates for sheep wool drop), seems to work (of course, this isn't a good test, but it shows that this PR is useful).

GoldenStack commented:
I reviewed this a few times but I honestly forgot what the blocker is. It might have been waiting on Matt having time and wanting to review it.

InfiniteCoder01 commented:
Just checked it on 26.1, works wonderfully.
Is there anything I can do to help this get merged? It seems like this didn't go anywhere since September...

kermandev committed:
Merge branch 'master' into fork/FluxCapacitor2/feat/component-predicates

kermandev committed:
fix: resolve compile and test errors after master merge

kermandev committed:
refactor: define range codecs in Range classes and register villager variant predicate

kermandev committed:
refactor: replace deprecated Check.notNull with Objects.requireNonNull and remove unused Check imports

kermandev committed:
refactor: design ComponentPredicateSet as sealed interface with ComponentPredicateSetImpl record and revert NetworkBuffer mapValue changes

kermandev committed:
fix: align component predicates with Minecraft 26.2

kermandev committed:
refactor: Remove NBT_COMPOUND_COERCED

kermandev committed:
fix: Bee

kermandev committed:
feat: Proper tooltip
